src/bson_mapper/test/archiver_test.cpp

```diff
@@ -0,0 +1,462 @@
+#include "catch.hpp"
+
+#include <bson_mapper/bson_archiver.hpp>
+#include <cereal/types/vector.hpp>
+#include <fstream>
+
+/**
+ * Returns true if the two files are identical, false otherwise.
+ */
+bool file_cmp(std::string filename1, std::string filename2) {
+    std::ifstream file1(filename1, std::ios_base::binary);
+    std::ifstream file2(filename2, std::ios_base::binary);
+
+    file1.seekg(0, file1.end);
+    file2.seekg(0, file2.end);
+
+    size_t file1_length = file1.tellg();
+    size_t file2_length = file2.tellg();
+
+    file1.seekg(0, file1.beg);
+    file2.seekg(0, file2.beg);
+
+    if (file1_length != file2_length) {
+        return false;
+    }
+
+    char file1buf[file1_length], file2buf[file2_length];
+
+    file1.read(file1buf, file1_length);
+    file2.read(file2buf, file2_length);
+
+    return std::memcmp(file1buf, file2buf, file1_length) == 0;
+}
+
+/**
+ * Counts the number of keys in a particular document view.
+ *
+ * @param v
+ *      the bsoncxx::document::view to count keys in.
+ *
+ * @return the number of keys in the view.
+ */
+size_t countKeys(bsoncxx::document::view v) {
+    size_t count = 0;
+    for (auto elem = v.begin(); elem != v.end(); ++elem) {
+        ++count;
+        // DEBUG: std::cout << " key: " << elem->key() << "\n";
+    }
+    return count;
+}
+
+struct DataA {
+    int32_t x, y;
+    double z;
+
+    template <class Archive>
+    void serialize(Archive& ar) {
+        ar(CEREAL_NVP(x), CEREAL_NVP(y), CEREAL_NVP(z));
+    }
+
+    bool operator==(const DataA& other) {
+        return x == other.x && y == other.y && z == other.z;
```

Comment:
There's a magical trick you can use with std::tuple

return std::tie(x, y, z) == std::tie(other.x, other.y, other.z);

works great when you have to do operator < as well